packages/cohere/src/cohere-chat-language-model.ts

```diff
@@ -304,3 +367,99 @@ const cohereChatChunkSchema = z.discriminatedUnion('event_type', [
     }),
   }),
 ]);
+
+// For reference: https://docs.cohere.com/docs/parameter-types-in-tool-use
+
+function prepareToolsAndToolChoice(
+  mode: Parameters<LanguageModelV1['doGenerate']>[0]['mode'] & {
+    type: 'regular';
+  },
+) {
+  const tools = mode.tools?.length ? mode.tools : undefined;
+
+  if (tools == null) {
+    return { tools: undefined };
+  }
+
+  const mappedTools = tools.map(tool => {
+    const { properties, required } = tool.parameters;
+
+    const parameterDefinitions: any = {};
+
+    if (properties) {
+      for (const [key, value] of Object.entries(properties)) {
+        if (typeof value === 'object' && value !== null) {
+          const { type: JSONType, description } = value;
+
+          let type: 'str' | 'float' | 'int' | 'bool';
+
+          if (typeof JSONType === 'string') {
+            switch (JSONType) {
+              case 'string':
+                type = 'str';
+                break;
+              case 'number':
+                type = 'float';
+                break;
+              case 'integer':
+                type = 'int';
+                break;
+              case 'boolean':
+                type = 'bool';
+                break;
+              default:
+                throw new Error(`Unsupported JSON type: ${JSONType}`);
```

Comment:
It would be good to use a dedicated AI SDK error here, e.g. `UnsupportedFunctionalityError`

Reply:
done!